XMonad/Util/Run.hs

```diff
@@ -180,3 +237,222 @@ spawnPipe' encoding x = io $ do
+--class calculator -e ghci"@.  The usefulness of this notation becomes
+apparent with more complicated examples:
+
+> proc $ inEmacs
+>    >-> withEmacsLibs [OwnFile "mailboxes"]
+>    >-> execute (elispFun "notmuch")
+>    >-> setFrameName "mail"
+
+This is equivalent to spawning
+
+> emacs -l /home/slot/.config/emacs/lisp/mailboxes.el
+>       -e '(notmuch)'
+>       -F '(quote (name . "mail"))'
+
+Notice how we did not have to specify the whole path to @mailboxes.el@,
+since we had set the correct 'emacsLispDir' upon starting xmonad.  This
+becomes especially relevant when running Emacs in batch mode, where one
+has to include [M,Non-GNU]ELPA packages in the call, whose exact names
+may change at any time.  Then the following
+
+> do url <- getSelection  -- from XMonad.Util.XSelection
+>    proc $ inEmacs
+>       >-> withEmacsLibs [ElpaLib "dash", ElpaLib "s", OwnFile "arXiv-citation"]
+>       >-> asBatch
+>       >-> execute (elispFun $ "arXiv-citation" <> asString url)
+
+becomes
+
+> emacs -L /home/slot/.config/emacs/elpa/dash-20220417.2250
+>       -L /home/slot/.config/emacs/elpa/s-20210616.619
+>       -l /home/slot/.config/emacs/lisp/arXiv-citation.el
+>       --batch
+>       -e '(arXiv-citation "<url-in-the-primary-selection>")'
+
+which would be quite bothersome to type indeed!
+
+-}
+
+-----------------------------------------------------------------------
+-- Types and whatnot
+
+-- | Additional information that might be useful when spawning external
+-- programs.
+data ProcessConfig = ProcessConfig
+  { editor :: String
+    -- ^ Default editor.  Defaults to @"emacsclient -c -a ''"@.
+  , emacsLispDir :: FilePath
+    -- ^ Directory for your custom Emacs lisp files.  Probably
+    -- @user-emacs-directory@ or @user-emacs-directory/lisp@.  Defaults
+    -- to @"~\/.config\/emacs\/lisp\/"@
+  , emacsElpaDir :: FilePath
+    -- ^ Directory for all packages from [M,Non-GNU]ELPA; probably
+    -- @user-emacs-directory/elpa@.  Defaults to
+    -- @"~\/.config\/emacs\/elpa"@.
+  , emacs :: String
+    -- ^ /Standalone/ Emacs executable; this should not be @emacsclient@
+    -- since, for example, the client does not support @--batch@ mode.
+    -- Defaults to @"emacs"@.
+  }
+
+-- | Given a 'ProcessConfig', remember it for spawning external
+-- processes later on.
+spawnExternalProcess :: ProcessConfig -> XConfig l -> XConfig l
+spawnExternalProcess = XC.modifyDef . const
+
+instance Default ProcessConfig where
+  def :: ProcessConfig
+  def = ProcessConfig
+    { editor       = "emacsclient -c -a ''"
+    , emacsLispDir = "~/.config/emacs/lisp/"
+    , emacsElpaDir = "~/.config/emacs/elpa/"
+    , emacs        = "emacs"
+    }
+
+-- | Convenient type alias.
+type Input = ShowS
+
+-----------------------------------------------------------------------
+-- Combinators
+
+-- | Combine inputs together.
+(>->) :: X Input -> X Input -> X Input
+(>->) = liftA2 (.)
+infixr 3 >->
```

Comment:
I know that this is the `Semigroup` instance for `X`, but I thought having a pipe symbol would be cute.  Not entirely sure about this though, since everywhere else we're composing things with `<>` (or `>>`).